packages/api-format/src/codec/Metadata.ts

```diff
@@ -0,0 +1,290 @@
+// Copyright 2017-2018 @polkadot/api-format authors & contributors
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+import CodecArray from './base/Array';
+import CodecEnum from './base/Enum';
+import CodecEnumType from './base/EnumType';
+import CodecOption from './base/Option';
+import CodecStruct from './base/Struct';
+import String from './String';
+import U16 from './U16';
+
+class EventMetadata extends CodecStruct<{
+  name: String,
+  arguments: CodecArray<String>,
+  documentation: CodecArray<String>
+}> {
+  constructor () {
```

Comment:
I really tried to not have to pass through the type and a structure that defines the classes (that will be used upon construction). I failed at my first attempt. So there is room for improvement, since what gets passed in the constructor is the same (more-or-less) as the generic type.

Reply:
Somewhere there must be a solution between key and lookup types, to not pass the type in, but just the struct in the constructor - https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#keyof-and-lookup-types